src/algebra/homology/homological_complex.lean

```diff
@@ -0,0 +1,646 @@
+by simp [C.d_to_eq r]
+
+@[simp, reassoc] lemma X_prev_iso_zero_comp_d_to {j : ι} (h : c.prev j = none) :
+  (C.X_prev_iso_zero h).inv ≫ C.d_to j = 0 :=
+by simp [h]
+
+@[simp, reassoc] lemma d_from_comp_X_next_iso {i j : ι} (r : c.rel i j) :
+  C.d_from i ≫ (C.X_next_iso r).hom = C.d i j :=
+by simp [C.d_from_eq r]
+
+@[simp, reassoc] lemma d_from_comp_X_next_iso_zero {i : ι} (h : c.next i = none) :
+  C.d_from i ≫ (C.X_next_iso_zero h).hom = 0 :=
+by simp [h]
+
+@[simp]
+lemma d_to_comp_d_from (j : ι) : C.d_to j ≫ C.d_from j = 0 :=
+begin
+  rcases h₁ : c.next j with _ | ⟨k,w₁⟩,
+  { rw [d_from_eq_zero _ h₁], simp, },
+  { rw [d_from_eq _ w₁],
+    rcases h₂ : c.prev j with _ | ⟨i,w₂⟩,
+    { rw [d_to_eq_zero _ h₂], simp, },
+    { rw [d_to_eq _ w₂], simp, } }
+end
+
+lemma kernel_from_eq_kernel [has_kernels V] {i j : ι} (r : c.rel i j) :
+  kernel_subobject (C.d_from i) = kernel_subobject (C.d i j) :=
+begin
+  rw C.d_from_eq r,
+  apply kernel_subobject_comp_mono,
+end
+
+lemma image_to_eq_image [has_images V] [has_equalizers V]
+  {i j : ι} (r : c.rel i j) :
+  image_subobject (C.d_to j) = image_subobject (C.d i j) :=
+begin
+  rw C.d_to_eq r,
+  apply image_subobject_iso_comp,
+end
+
+end
+
+namespace hom
+
+variables {C₁ C₂ C₃ : homological_complex V c}
+variables [has_zero_object V]
+local attribute [instance] has_zero_object.has_zero
+
+/-! Lemmas relating chain maps and `d_to`/`d_from`. -/
+
+/-- `f.f_prev j` is `f.f i` if there is some `r i j`, and zero otherwise. -/
+def f_prev (f : hom C₁ C₂) (j : ι) : C₁.X_prev j ⟶ C₂.X_prev j :=
+match c.prev j with
+| none := 0
+| some ⟨i,w⟩ := (C₁.X_prev_iso w).hom ≫ f.f i ≫ (C₂.X_prev_iso w).inv
+end
+
+lemma f_prev_eq (f : hom C₁ C₂) {i j : ι} (w : c.rel i j) :
+  f.f_prev j = (C₁.X_prev_iso w).hom ≫ f.f i ≫ (C₂.X_prev_iso w).inv :=
+begin
+  dsimp [f_prev],
+  rw c.prev_eq_some w,
+  refl,
+end
+
+/-- `f.f_next i` is `f.f j` if there is some `r i j`, and zero otherwise. -/
+def f_next (f : hom C₁ C₂) (i : ι) : C₁.X_next i ⟶ C₂.X_next i :=
+match c.next i with
+| none := 0
+| some ⟨j,w⟩ := (C₁.X_next_iso w).hom ≫ f.f j ≫ (C₂.X_next_iso w).inv
+end
+
+lemma f_next_eq (f : hom C₁ C₂) {i j : ι} (w : c.rel i j) :
+  f.f_next i = (C₁.X_next_iso w).hom ≫ f.f j ≫ (C₂.X_next_iso w).inv :=
+begin
+  dsimp [f_next],
+  rw c.next_eq_some w,
+  refl,
+end
+
+@[simp, reassoc]
+lemma comm_from (f : hom C₁ C₂) (i : ι) :
+  f.f i ≫ C₂.d_from i = C₁.d_from i ≫ f.f_next i :=
+begin
+  rcases h : c.next i with _ | ⟨j,w⟩,
+  { simp [h], },
+  { simp [d_from_eq _ w, f_next_eq _ w], }
+end
+
+@[simp, reassoc]
+lemma comm_to (f : hom C₁ C₂) (j : ι) :
+  f.f_prev j ≫ C₂.d_to j = C₁.d_to j ≫ f.f j :=
+begin
+  rcases h : c.prev j with _ | ⟨j,w⟩,
+  { simp [h], },
+  { simp [d_to_eq _ w, f_prev_eq _ w], }
+end
+
+/--
+A morphism of chain complexes
+induces a morphism of arrows of the differentials out of each object.
+-/
+def sq_from (f : hom C₁ C₂) (i : ι) : arrow.mk (C₁.d_from i) ⟶ arrow.mk (C₂.d_from i) :=
```

Comment:
One could alternatively consider a morphism of (co)chain complexes as a (co)chain complex of arrows, and whenever you have a square (i.e. a morphism of arrows), it can be "reflected" to obtain another square (I don't know whether "reflected" is an actual term that's used). This (and `sq_to`) end up being the reflections of the differentials in the associated (co)chain complex of arrows. I don't know whether this point of view is useful at all, or if it simplifies some things. Just figured I would mention it in case you think it might be helpful...

Reply:
I think we can experiment with this in a later PR. But I think the sort of explicit hands-on version will probably be very useful to have.